### PR TITLE
Add mock useradd to pwd

### PR DESCRIPTION
**Pull request by maorJsonar**

### CODE REVIEW

File: pwd.py [hunk 1]
```diff
@@ -30,3 +32,18 @@ def getpwnam(self, name):
...
+                raise TypeError(f"expected str, bytes or os.PathLike object, not {type(arg)}")
+        for user in self.db:
+            if uid is not None and int(uid) == user.pw_uid:
+                raise CalledProcessError(f"invalid user ID 'id'")
+            if username == user.pw_name:
+                raise CalledProcessError(f"useradd: user '{username}' already exists")
+        if uid is None:
+            uid = max([user.pw_uid for user in self.db]) + 1
+
+        self.db.append(struct_passwd(username, 'x', int(uid), 981,
```
Comment (oorijsonar) [inline pwd.py:48]: If you got a `groupname` then shouldn't the `gid` correspond to it and not be a hard coded 981? Does the other mock you wrote have the same problem?

File: pwd.py [hunk 2]
```diff
@@ -30,3 +32,18 @@ def getpwnam(self, name):
 
     def get_pwall(self):
         return self.db
+
+    def useradd(self, username, uid, groupname):
+        for arg in (username, groupname, uid):
+            if arg is not None and not isinstance(arg, (str, bytes, os.PathLike)):
+                raise TypeError(f"expected str, bytes or os.PathLike object, not {type(arg)}")
```
Comment (oorijsonar) [inline pwd.py:39]: This code looks familiar...

**Review by oorijsonar — CHANGES_REQUESTED** [review]

File: pwd.py (hunk 2 above)
Comment (maorJsonar) [reply pwd.py]: Added base class with this code.

File: pwd.py (hunk 1 above)
Comment (maorJsonar) [reply pwd.py:47]: We have decided to leave this as is for now.

**Comment by yhagerjs** [comment]
Why do we need this? The original grp/pwd don't have this function. What's the point in mocking something that doesn't exist? The mocked classes already contain a mechanism to initialize a list of users/groups - the idea was to use that and keep the original API, not enhance the stdlib grp/pwd modules.

**Comment by oorijsonar** [comment]
> Why do we need this? The original grp/pwd don't have this function. What's the point in mocking something that doesn't exist? The mocked classes already contain a mechanism to initialize a list of users/groups - the idea was to use that and keep the original API, not enhance the stdlib grp/pwd modules.

To be able to test that we pass the arguments properly (we didn't pass them properly https://github.com/jsonar/sonarg/issues/1776).

**Comment by yhagerjs** [comment]
You are referring to the string validation? I'm asking about the `useradd`/`groupadd` functions.

**Comment by oorijsonar** [comment]
> You are referring to the string validation? I'm asking about the `useradd`/`groupadd` functions.

Please see the pull request that uses this one.

File: ent.py [hunk 3]
```diff
@@ -0,0 +1,10 @@
+import os
+
+
+class MockEnt:
```
Comment (oorijsonar) [inline ent.py:4]: 1. Rename this file to `subprocess.py`
2. Rename the function to `validate` or `validate_run_args`.
3. Make the function the only thing in this file (no classes).

**Review by oorijsonar — CHANGES_REQUESTED** [review]

File: grp.py [hunk 4]
```diff
@@ -1,12 +1,13 @@
...
 from subprocess import CalledProcessError
 
+from test.mock.subprocess import validate_run_args
+
 struct_group = namedtuple('struct_group',
                           ['gr_name', 'gr_passwd', 'gr_gid', 'gr_mem'])
 
 
-class MockGrp:
+class MockGrp():
```
Comment (oorijsonar) [inline grp.py:10]: No need for this change.

File: pwd.py [hunk 5]
```diff
@@ -1,11 +1,14 @@
...
+
+from test.mock.subprocess import validate_run_args
 
 struct_passwd = namedtuple('struct_passwd',
                            ['pw_name', 'pw_passwd', 'pw_uid',
                             'pw_gid', 'pw_gecos', 'pw_dir', 'pw_shell'])
 
 
-class MockPwd:
+class MockPwd():
```
Comment (oorijsonar) [inline pwd.py:11]: No need for this change.

**Review by oorijsonar — CHANGES_REQUESTED** [review]

**Pull request merged**